single_kernel_mongo/state/charm_state.py

```diff
@@ -0,0 +1,791 @@
+        return self.secrets.set(AppPeerDataKeys.KEYFILE.value, keyfile_content, Scope.APP).label
+
+    def get_keyfile(self) -> str | None:
+        """Gets the keyfile content from the secret."""
+        return self.secrets.get_for_key(Scope.APP, AppPeerDataKeys.KEYFILE.value)
+
+    @property
+    def planned_units(self) -> int:
+        """Return the planned units for the charm."""
+        return self.model.app.planned_units()
+
+    @cached_property
+    def peer_units_data_interfaces(self) -> dict[Unit, DataPeerOtherUnitData]:
+        """The cluster peer relation."""
+        return {
+            unit: DataPeerOtherUnitData(
+                model=self.model, unit=unit, relation_name=self.peer_relation_name
+            )
+            for unit in self.peers_units
+        }
+
+    @cached_property
+    def upgrade_units_data_interfaces(self) -> dict[Unit, DataPeerOtherUnitData]:
+        """The cluster peer relation."""
+        return {
+            unit: DataPeerOtherUnitData(
+                model=self.model, unit=unit, relation_name=RelationNames.UPGRADE_VERSION.value
+            )
+            for unit in self.peers_units
+        }
+
+    @property
+    def formatted_socket_path(self) -> str:
+        """URL encoded socket path.
+
+        Explanation: On Mongos VM which is a subordinate charm, we'd rather
+        share the connection with a socket in order to improve latency.
+        """
+        return quote(f"{self.paths.socket_path}", safe="")
+
+    @property
+    def _socket_path(self) -> set[str] | None:
+        """If we prefer to use a socket, return it.
+
+        Otherwise return None.
+        """
+        if (
+            self.substrate == Substrates.VM
+            and self.charm_role.name == CharmKind.MONGOS
+            and not self.app_peer_data.external_connectivity
+        ):
+            return {self.formatted_socket_path}
+        return None
+
+    @property
+    def app_hosts(self) -> set[str]:
+        """Retrieve the hosts associated with MongoDB application."""
+        return self._socket_path or {unit.host for unit in self.units}
+
+    @property
+    def internal_hosts(self) -> set[str]:
+        """Internal hosts for internal access."""
+        return self._socket_path or {unit.internal_address for unit in self.units}
+
+    @property
+    def host_port(self) -> int:
+        """Retrieve the port associated with MongoDB application."""
+        if self.is_role(MongoDBRoles.MONGOS):
+            if self.app_peer_data.external_connectivity:
+                return self.unit_peer_data.node_port
+            return MongoPorts.MONGOS_PORT
+        return MongoPorts.MONGODB_PORT
+
+    @property
+    def config_server_data_interface(self) -> DatabaseProviderData:
+        """The config server database interface."""
+        return DatabaseProviderData(self.model, RelationNames.CONFIG_SERVER.value)
+
+    @property
+    def shard_state_interface(self) -> DatabaseRequirerData:
```

Comment:
i notice that you have a `state/config_server.py` file. Does it make sense to also have a `state/shard.py` file?

Reply:
They use the same fields (because they are two sides of the same coin), that's why I used a single file.
I'm really not fully satisfied with the implementation of the interfaces databag for now in this new implementation and I'll try to do better but it might need some rework of the data interfaces so it might be for a bit later :)

Reply:
no worries- thank you for the explaination